README.md
# udpipe-hs
Basic haskell bindings for [udpipe](https://github.com/ufal/udpipe)

## Example

This code

```haskell
runExample = do
    let modelName = "models/english-lines-ud-2.5-191206.udpipe"
    mod <- loadModel modelName
    case mod of
      Left msg -> putStrLn msg
      Right mod -> do
        let input = "This is an example"
        ans <- runPipeline mod input
        putStrLn ans
```

will produce this output

```
# newdoc
# newpar
# sent_id = 1
# text = This is an example
1       This    this    PRON    DEM-SG  Number=Sing|PronType=Dem        4       nsubj   _       _
2       is      be      AUX     PRES    Mood=Ind|Number=Sing|Person=3|Tense=Pres|VerbForm=Fin   4       cop     _       _
3       an      an      DET     IND-SG  Definite=Ind|PronType=Art       4       det     _       _
4       example example NOUN    SG-NOM  Number=Sing     0       root    _       SpacesAfter=\n
```

assuming you have downloaded `english-lines-ud-2.5-191206.udpipe` and put it in the `models` directory.

You can download models here: https://ufal.mff.cuni.cz/udpipe/1/models


CHANGELOG.md
# Revision history for udpipe-hs

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


src/MyLib.hs
{- |
A very simple binding for udpipe.

Example usage:

> import UDPipe
>
> main = do
>   let sentence = "The quick brown fox jumps over the lazy dog."
>   let modelFile = "models/english-lines-ud-2.5-191206.udpipe"
>   model' <- loadModel modelFile
>   let model = case model' of
>                 Left err -> error err
>                 Right m -> m
>   result <- runPipe model sentence
>   putStrLn result


-}
module MyLib (runExample, runPipeline, loadModel, ModelPtr) where
import Foreign
import Foreign.C

-- | A loaded udpipe model.
type ModelPtr = Ptr Model

data Model

foreign import ccall unsafe "load_model" cLoadModel
  :: CString -> IO ModelPtr

foreign import ccall unsafe "run_pipeline_simple" cRunPipeline
  :: ModelPtr -> CString -> IO CString

foreign import ccall unsafe "free_cstr" freeCString
  :: CString -> IO ()

-- | Run the default ud pipeline on an input string.
runPipeline :: ModelPtr -> String -> IO String
runPipeline model input = do
  withCString input $ \cInput -> do
    cOutput <- cRunPipeline model cInput
    output <- peekCString cOutput
    freeCString cOutput
    return output

-- | Load a model from a file.
loadModel :: FilePath -> IO (Either String ModelPtr)
loadModel path = do
  withCString path $ \cPath -> do
    model <- cLoadModel cPath
    if model == nullPtr
      then return $ Left $ "Failed to load model from " ++ path
      else return $ Right model

-- | An example of how to use the library.
runExample :: IO ()
runExample = do
    let modelName = "models/english-lines-ud-2.5-191206.udpipe"
    mod <- loadModel modelName
    case mod of
      Left msg -> putStrLn msg
      Right mod -> do
        let input = "This is an example"
        ans <- runPipeline mod input
        putStrLn ans

-- TODO: Handle errors in the C++ code (give back error message)


cbits/capi.cpp
#include "common.h"
// #include "model/evaluator.h"
#include "model/model.h"
#include "model/pipeline.h"
// #include "sentence/input_format.h"
// #include "sentence/output_format.h"
// #include "tokenizer/detokenizer.h"
// #include "trainer/trainer.h"
// #include "utils/getpara.h"
// #include "utils/iostreams.h"
// #include "utils/options.h"
// #include "utils/process_args.h"
// #include "version/version.h"
#include "capi.h"

#include <sstream>
#include <iostream>

using namespace ufal::udpipe;

extern "C" {

model* load_model(char* filename) {
  return model::load(filename);
}

void free_model(model* m) {
  // std::cerr << "Releasing model" << std::endl;
  delete m;
}

// pipeline* make_pipeline(model* m, char* input, char* tagger, char* parser, char* output) {
//   auto pl = new pipeline(m, input, tagger, parser, output);
//   return pl;
//   return reinterpret_cast<pipeline*>(pl);
// }

bool run_pipeline(model* m, const string& input_type, const string& tagger, const string& parser, const string& output_type, 
                    std::istream& input_stream, std::ostream& output_stream, string& error) {
  pipeline pipeline(m, input_type, tagger, parser, output_type);
  return pipeline.process(input_stream, output_stream, error);
}

const char* run_pipeline_simple(model* m, char* input, bool* result) {
  std::stringstream instream(input);
  std::ostringstream outstream;
  string error;
  *result = run_pipeline(m, "tokenize", pipeline::DEFAULT, pipeline::DEFAULT, "conllu", instream, outstream, error);
  // cerr << "Success: " << *result << endl;
  // cerr << "Error: " << error << endl;
  string str = outstream.str();
  // cerr << "Result: " << str << endl;
  if (*result) {
    return strdup(str.c_str());
  } else {
    return strdup(error.c_str());
  }
  // char * cstr = new char [str.length()+1];
  // std::strcpy (cstr, str.c_str());
  // return cstr;
}

void free_cstr(char* cstr) {
  delete[] cstr;
}


}


src/UDPipe.hs
{- |
A very simple binding for udpipe.

Example usage:

> import UDPipe
>
> main = do
>   let sentence = "The quick brown fox jumps over the lazy dog."
>   let modelFile = "models/english-lines-ud-2.5-191206.udpipe"
>   model' <- loadModel modelFile
>   let model = case model' of
>                 Left err -> error err
>                 Right m -> m
>   result <- runPipe model sentence
>   putStrLn result


-}
module UDPipe (runExample, runPipeline, loadModel, Model) where
import Foreign
    ( Ptr, FunPtr, nullPtr, newForeignPtr, withForeignPtr, ForeignPtr, Storable (peek) )
import Foreign.C ( peekCString, withCString, CString, CBool )
import Control.Exception (bracket, mask_)
import Foreign.Marshal.Alloc (alloca)

-- | A loaded udpipe model.
newtype Model = Model (ForeignPtr ModelRep)
  deriving (Show, Eq, Ord)

type ModelPtr = Ptr ModelRep

data ModelRep

foreign import ccall unsafe "load_model" cLoadModel
  :: CString -> IO ModelPtr

foreign import ccall unsafe "run_pipeline_simple" cRunPipeline
  :: ModelPtr -> CString -> Ptr CBool -> IO CString

foreign import ccall unsafe "free_cstr" freeCString
  :: CString -> IO ()

foreign import ccall unsafe "&free_model" freeModel
  :: FunPtr (ModelPtr -> IO ())

cBoolToBool :: CBool -> Bool
cBoolToBool = toEnum . fromEnum

-- | Run the default ud pipeline on an input string.
runPipeline :: Model -> String -> IO (Either String String)
runPipeline (Model fmodel) input = do
  withCString input $ \cInput ->
    alloca $ \successPtr ->
      withForeignPtr fmodel $ \model ->
        bracket (cRunPipeline model cInput successPtr) freeCString $ \ x -> do
          success <- peek successPtr
          str <- peekCString x
          return $ if cBoolToBool success then Right str else Left str

-- | Load a model from a file.
loadModel :: FilePath -> IO (Either String Model)
loadModel path = do
  withCString path $ \cPath -> mask_ $ do
    -- mask_ is needed to avoid leaking the pointer in case an async exception
    -- occurs between allocation and wrapping it in a foreign pointer.
    model <- cLoadModel cPath
    if model == nullPtr
      then return $ Left $ "Failed to load model from " ++ path
      else do
        fmodel <- newForeignPtr freeModel model
        return $ Right $ Model fmodel

-- | An example of how to use the library.
runExample :: IO ()
runExample = do
    let modelName = "models/english-lines-ud-2.5-191206.udpipe"
    mod <- loadModel modelName
    case mod of
      Left msg -> putStrLn msg
      Right mod -> do
        let input = "This is an example"
        ans <- runPipeline mod input
        putStrLn $ either ("Error running pipeline: "++) id ans

{- TODO: 
 - * Handle errors in the C++ code (give back error message)
 -} 